test/integration/client/client_test.go

```diff
@@ -1342,3 +1347,59 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
 		t.Errorf("createMgrExtracted apply configuration did not match expected, got:\n%s\n", cmp.Diff(expectedCreateExtracted, createMgrExtracted))
 	}
 }
+
+func TestAPIServerTransportMetrics(t *testing.T) {
+	result := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+	defer result.TearDownFn()
+
+	c := clientset.NewForConfigOrDie(result.ClientConfig).RESTClient()
+
+	body, err := c.Get().AbsPath("/metrics").DoRaw(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO: this can be much better if there is some library that parse prometheus metrics
+	// the existing one in "k8s.io/component-base/metrics/testutil" uses the global variable
+	// but we want to parse the ones returned by the endpoint to be sure the metrics are
+	// exposed correctly
+	var hits, misses, entries int
+	for _, line := range strings.Split(string(body), "\n") {
+		if !strings.HasPrefix(line, "rest_client_transport") {
+			continue
+		}
+		if strings.Contains(line, "not-cacheable") {
+			t.Fatalf("detected transport that is not cacheable, please check https://issues.k8s.io/112017")
+		}
+
+		output := strings.Split(line, " ")
+		if len(output) != 2 {
+			t.Fatalf("expected metrics to be in the format name value, got %v", output)
+		}
+		name := output[0]
+		value, err := strconv.Atoi(output[1])
+		if err != nil {
+			t.Fatalf("metric value can not be converted to integer %v", err)
+		}
+		switch name {
+		case "rest_client_transport_cache_entries":
+			entries = value
+		case `rest_client_transport_create_calls_total{result="hit"}`:
+			hits = value
+		case `rest_client_transport_create_calls_total{result="miss"}`:
+			misses = value
+		}
+		t.Logf("metric %s", line)
+	}
+
+	if misses != entries || misses == 0 {
+		t.Errorf("expected as many entries %d in the cache as misses, got %d", entries, misses)
+	}
+
+	if hits < misses {
+		t.Errorf("expected more hits %d in the cache than misses %d", hits, misses)
+	}
+
+	// TODO: we can keep querying for some time to see if the misses keep growing over time
```

Comment:
Would it be valuable to spam ~1k updates to an API service and them assert that the total number of client has not increased more than say 100?  Basically a test that would fail if we reverted / regressed #117258 ?

Reply:
I like that

Reply:
done, aiming for a 90% hit ratio in the cache, that I think is reasonable